src/Evolution/Systems/GrMhd/ValenciaDivClean/NewmanHamlin.cpp

```diff
@@ -123,7 +123,11 @@ std::optional<PrimitiveRecoveryData> NewmanHamlin::apply(
                         Scalar<double>(current_rest_mass_density),
                         Scalar<double>(current_pressure)));
       } else if constexpr (ThermodynamicDim == 3) {
-        ERROR("3d EOS not implemented");
+        current_specific_internal_energy =
+            (rho_h_w_squared / square(current_lorentz_factor) -
+             current_pressure) /
+                current_rest_mass_density -
+            1.0;
```

Comment:
LGTM, this has a catastrophic cancellation, but I don't know what to do about that other than to rework the entire recovery scheme (not worth it!)

Reply:
Yeah, I would guess we would be using Kastaun most of the time anyway, if this becomes a serious limitation we  can take a look at trying to patch it up.

Reply:
Yea, I think the "solution" is "use a better recovery scheme" 😅 Kastaun isn't perfect in this regard, but it isn't obvious to me that a "perfect" scheme is even possible

Reply:
I think something better would probably go towards a physicality preserving evolution scheme, but that is even harder :)

Reply:
Yes, that would be the dream... But agreed, very, very difficult :(